Assets/scripts/Weapons/Weapon.cs

```diff
@@ -0,0 +1,265 @@
+		/// The amount of times per second this weapon can fire
+		/// </summary>
+		public double FireRate;
+		/// <summary>
+		/// The amount of projectiles expended per shot
+		/// </summary>
+		public int ProjectilesFired;
+		/// <summary>
+		/// The type of ammo this weapon will allow, this is a string and not an enum for diversity
+		/// </summary>
+		public string AmmoType;
+		/// <summary>
+		/// The current magazine for this weapon, null means empty
+		/// </summary>
+		private MagazineBehaviour CurrentMagazine;
+		/// <summary>
+		/// The countdown untill we can shoot again
+		/// </summary>
+		[HideInInspector]
+		public double FireCountDown;
+		/// <summary>
+		/// If the weapon is currently in automatic action
+		/// </summary>
+		[HideInInspector]
+		public bool InAutomaticAction;
+
+		[SyncVar(hook="LoadUnloadAmmo")]
+		public NetworkInstanceId MagNetID;
+
+		[SyncVar]
+		public string ControlledByPlayer;
+
+		void Start()
+		{
+			InAutomaticAction = false;
+
+			//init weapon with missing settings
+			if (AmmoType == null)
+				AmmoType = "12mm";
+			if (Projectile == null)
+				Projectile = Resources.Load("Bullet_12mm") as GameObject;
+		}
+
+		void Update () {
+			if (FireCountDown > 0 )
+			{
+				FireCountDown -= Time.deltaTime;
+				//prevents the next projectile taking miliseconds longer than it should
+				if (FireCountDown < 0) {
+					FireCountDown = 0;
+				}
+			}
+
+			if (Input.GetKeyDown (KeyCode.E)) {
+				ManualUnload(CurrentMagazine);
+			}
+
+			if(Input.GetMouseButtonUp(0)) {
+				InAutomaticAction = false;
+			}
+
+			if (InAutomaticAction && FireCountDown <= 0) {
+				AttemptToFireWeapon();
+			}
+		}
+
+		#region Weapon Server Init
+		public override void OnStartServer()
+		{
+			GameObject m = GameObject.Instantiate(Resources.Load("Magazine_"+AmmoType) as GameObject, Vector3.zero, Quaternion.identity);
+			NetworkServer.Spawn(m);
+			StartCoroutine(SetMagazineOnStart(m));
+			base.OnStartServer();
+		}
+
+		//Gives it a chance for weaponNetworkActions to init
+		IEnumerator SetMagazineOnStart(GameObject magazine){
+			yield return new WaitForEndOfFrame();
+			PlayerManager.LocalPlayerScript.weaponNetworkActions.CmdLoadMagazine(gameObject, magazine);
+		}
+		#endregion
+
+		#region Weapon Firing Mechanism
+		public override void Interact() {
+			//shoot gun interation if its in hand
+			if (gameObject == UIManager.Hands.CurrentSlot.GameObject()) {
+				AttemptToFireWeapon();
+			} 
+			//if the weapon is not in our hands not in hands, pick it up
+			else {
+				PlayerManager.LocalPlayerScript.playerNetworkActions.TryToPickUpObject(gameObject);
+			}
+		}
+
+		void AttemptToFireWeapon()
+		{
+			//ignore if we are hovering over UI
+			if (UnityEngine.EventSystems.EventSystem.current.IsPointerOverGameObject())
+				return;
+
+			//if the hand with the weapon in it is selected
+			if (UIManager.Hands.CurrentSlot == UIManager.Hands.LeftSlot || UIManager.Hands.CurrentSlot == UIManager.Hands.RightSlot) {
```

Comment:
If there's a check for the FireCountDown on every path except the first one, can we wrap the rest of the paths in one 
`if (FireCoutnDown <= 0`
Means it doesn't bother checking the other clauses every time